DOCSP-41121 Connect to Mongo Client (#34)
mongodb/docs-kotlin-sync: source/connect/mongoclient.txt

.. _kotlin-sync-mongoclient:

====================
Create a MongoClient
====================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: connection string, URI, server, Atlas, settings, client

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

To connect to a MongoDB deployment, you need two things:

- A **connection URI**, also known as a *connection string*, which tells the {+driver-short+}
  which MongoDB deployment to connect to.

> - A **MongoClient** object, which creates the connection to and perform 

Reviewer: perform -> performs

  operations on the MongoDB deployment.

> You can also use either of these components to customize the way the {+driver-short+} behaves

Reviewer: You can't really use MongoClient to customize.  You can use MongoClientSettings though.

while connected to MongoDB.

This guide shows you how to create a connection string and use a ``MongoClient`` object
to connect to MongoDB.

.. _kotlin-sync-connection-uri:

Connection URI
--------------

A standard connection string includes the following components:

> .. TODO, add this as last sentence for ``username:password`` description: For more information about the ``authSource`` 

Reviewer: Does this TODO need to be done?

Author: A reminder for when the kotlin authentication page is created. Made comment clearer, thanks!

.. connection option, see :ref:`kotlin-sync-auth`. 

> .. list-table::

Reviewer: Are these all a direct copy from Java driver?  Do we want to duplicate it like this?  Up to you but seems hard to maintain.

Author: Separate sites are maintained for each driver as of right now. For the Kotlin Sync work, we are using a combination of the [PyMongo](https://www.mongodb.com/docs/languages/python/pymongo-driver/current/connect/mongoclient/) and [Kotlin Coroutine](https://www.mongodb.com/docs/drivers/kotlin/coroutine/current/fundamentals/connection/connect/) docs as a model.

   :widths: 20 80
   :header-rows: 1

   * - Component
     - Description

   * - ``mongodb://``

     - Required. A prefix that identifies this as a string in the
       standard connection format.

   * - ``username:password``

     - Optional. Authentication credentials. If you include these, the client
       authenticates the user against the database specified in ``authSource``.

   * - ``host[:port]``

     - Required. The host and optional port number where MongoDB is running. If you don't
       include the port number, the driver uses the default port, ``27017``.

   * - ``/defaultauthdb``
     
     - Optional. The authentication database to use if the
       connection string includes ``username:password@``
       authentication credentials but not the ``authSource`` option. If you don't include 
       this component, the client authenticates the user against the ``admin`` database.

   * - ``?<options>``

     - Optional. A query string that specifies connection-specific
       options as ``<name>=<value>`` pairs. See

>        :ref:`kotlin-sync-connection-options` for a full description of

Reviewer: Is this page already live?  If not, when will it be added?

Author: As of right now, this links to a placeholder page for connection options. The page should be done in another few weeks.

       these options.

For more information about creating a connection string, see
:manual:`Connection Strings </reference/connection-string?tck=docs_driver_kotlin>` in the
MongoDB Server documentation.

Atlas Connection Example
------------------------

To connect to a MongoDB deployment on Atlas, you must first create a client. 
You can create a client with your desired configurations by passing a 
``MongoClientSettings`` object to the ``MongoClient.create()`` method.

> To instantiate a ``MongoClientSettings`` object, use the builder method to 

Reviewer: Should we also say that you can create the MongoClient directly from a connection string?

Author: Added!

   :header-rows: 1

   * - Component
     - Description

   * - ``mongodb://``

     - Required. A prefix that identifies this as a string in the
       standard connection format.

   * - ``username:password``

     - Optional. Authentication credentials. If you include these, the client
       authenticates the user against the database specified in ``authSource``.

   * - ``host[:port]``

     - Required. The host and optional port number where MongoDB is running. If you don't
       include the port number, the driver uses the default port, ``27017``.

   * - ``/defaultauthdb``
     
     - Optional. The authentication database to use if the
       connection string includes ``username:password@``
       authentication credentials but not the ``authSource`` option. If you don't include 
       this component, the client authenticates the user against the ``admin`` database.

   * - ``?<options>``

     - Optional. A query string that specifies connection-specific
       options as ``<name>=<value>`` pairs. See
       :ref:`kotlin-sync-connection-options` for a full description of
       these options.

For more information about creating a connection string, see

> :manual:`Connection Strings </reference/connection-string?tck=docs_driver_java_sync>` in the

Author: ```suggestion
:manual:`Connection Strings </reference/connection-string?tck=docs_driver_java_sync>` in the
```
Is the ``tck`` correct for the java sync language in the selector? @mcmorisi

Reviewer 2: Java Sync is probably the closest match but it's still technically a different language (despite sharing a lot of syntax with Kotlin). I see that there's no option in the dropdown for kotlin, so I think we can remove the `?tck=docs_driver_java_sync`

     - Description

   * - ``mongodb://``

     - Required. A prefix that identifies this as a string in the
       standard connection format.

   * - ``username:password``

     - Optional. Authentication credentials. If you include these, the client
       authenticates the user against the database specified in ``authSource``.
       For more information about the ``authSource`` connection option, see
       :ref:`pymongo-auth`. 

   * - ``host[:port]``

     - Required. The host and optional port number where MongoDB is running. If you don't
       include the port number, the driver uses the default port, ``27017``.

   * - ``/defaultauthdb``
     
     - Optional. The authentication database to use if the
       connection string includes ``username:password@``
       authentication credentials but not the ``authSource`` option. If you don't include 
       this component, the client authenticates the user against the ``admin`` database.

   * - ``?<options>``

     - Optional. A query string that specifies connection-specific
       options as ``<name>=<value>`` pairs. See
       :ref:`kotlin-sync-connection-options` for a full description of
       these options.

> .. Is there relevant connection string content for kotlin sync to ref to? Only see for java sync.

Reviewer 2: I would confirm with the Java DBX team but it's usually the case that Java and Kotlin content overlaps heavily.

When in doubt, we can always link readers to the [Server manual](https://www.mongodb.com/docs/manual/reference/connection-string/) for more general-purpose reading on a subject.

```suggestion
```

Author: Hi @jyemin, do you know if the connection string information for java sync is applicable for kotlin sync? I would like to link to connection string information on this page. This is the connection string page for java sync: https://www.mongodb.com/docs/manual/reference/connection-string/?tck=docs_driver_python

Reviewer: Yes, connection string info is exactly the same.

For more information about creating a connection string, see
:manual:`Connection Strings </reference/connection-string?tck=docs_driver_kotlin>` in the
MongoDB Server documentation.

Atlas Connection Example
------------------------

To connect to a MongoDB deployment on Atlas, create a client. You can create a 
client that uses your connection string and other client options by passing a 

> ``MongoClientSettings`` object to the ``MongoClient.create()`` method.

Author: Good revision, thanks!

> To instantiate a ``MongoClientSettings`` object, use the builder method to 
> specify your connection string and any other client options, and then call 
> the ``build()`` method. Chain the ``applyConnectionString()``` method to the 
> builder to specify your connection URI.

Reviewer 2: ```suggestion
To instantiate a ``MongoClientSettings`` object, use the builder method to 
specify your connection string and any other client options by chaining the appropriate methods.
Once you have set your desired configurations, call the ``build()`` method.
```

Author: Revised!

You can set the Stable API version client option to avoid breaking changes when 
you upgrade to a new server version. To learn more about the Stable API feature, 
see the :ref:`Stable API page <kotlin-sync-stable-api>`. 

The following code shows how you can specify the connection string and the 
Stable API client option when connecting to a MongoDB deployment on Atlas 
and verify that the connection is successful:

.. literalinclude:: /includes/connect/mongoclient.kt
   :start-after: start-connect-to-atlas
   :end-before: end-connect-to-atlas
   :language: kotlin
   :copyable:
   :dedent:

> Connect to a MongoDB Server or Replica Set

Reviewer 2: I think everything in this section can be removed – it overlaps with the [Choose a Connection Target](https://github.com/mongodb/docs-kotlin-sync/pull/33) page currently in PR. I think this page should strictly cover creating a client, while the Choose a Connection Target page contains details about the different types of MongoDB instances you can connect to.
